### PR TITLE
Adding MulticastAppDelegate

### PR DESCRIPTION
**Pull request by chliangGoogle**
This will be a feature branch after this PR approval, this PR just mimics the oiginal implementation: https://github.com/google/GoogleUtilities/pull/50.

I will check out subbranch from here to continue work on SceneDelegate support for auth and UNUserNotificationCenterDelegate for messaging

### CODE REVIEW

File: GoogleMulticastAppDelegate.podspec
Comment (paulb777) [inline GoogleMulticastAppDelegate.podspec:30]: Delete now that this is Objective C

File: GoogleMulticastAppDelegate.podspec
Comment (paulb777) [inline GoogleMulticastAppDelegate.podspec:2]: Is this useful for non-Firebase libraries? If no, how about FirebaseMulticastAppDelegate. If yes, it should be GoogleUtilitiesMulticastAppDelegate to avoid expanding our `Google` namespace.

File: GoogleMulticastAppDelegate.podspec
Comment (paulb777) [inline GoogleMulticastAppDelegate.podspec:4]: Open question for discussion. Should the versioning match GoogleUtilities? For SPM, should this be part of the GoogleUtilities package?

File: GoogleMulticastAppDelegate/Apps/SwiftUISample/SwiftUISample/MulticastAppDelegate.swift
Comment (paulb777) [inline GoogleMulticastAppDelegate/Apps/SwiftUISample/SwiftUISample/MulticastAppDelegate.swift:1]: Update copyrights

File: GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m
Comment (paulb777) [inline GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m:97]: Why not macos?

**Review by paulb777 — COMMENTED** [review]
Do you plan to add unit tests and CI?

File: GoogleMulticastAppDelegate/Sources/Public/GoogleUtilities/GULMulticastAppDelegate.h
Comment (chliangGoogle) [inline GoogleMulticastAppDelegate/Sources/Public/GoogleUtilities/GULMulticastAppDelegate.h:21]: @paulb777 temporarily copy over GULApplication.h from AppDelegateSwizzler. I wonder if it's possible to create a public header folder depended by GoogleUtilities/AppDelegateSwizzler and GULMulticastAppDelegate.

Comment (paulb777) [reply GoogleMulticastAppDelegate/Sources/Public/GoogleUtilities/GULMulticastAppDelegate.h:21]: It might be cleaner to make a GULMApplication.h and then we can cleanly delete everything about the old swizzler in a breaking change update.

Comment (chliangGoogle) [reply GoogleMulticastAppDelegate/Sources/Public/GoogleUtilities/GULMulticastAppDelegate.h]: Discussed offline, will let multicastAppDelegate depend on AppDelegateSwizzler since all users are using swizzling. Will move common code to multicastAppDelegate once we deprecate and remove swizzling.

File: GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m
Comment (chliangGoogle) [reply GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m:130]: We just always use TARGET_OS for the same swizzling logic. I'm not sure what's the tradeoff here.

File: GoogleMulticastAppDelegate.podspec
Comment (chliangGoogle) [reply GoogleMulticastAppDelegate.podspec:30]: After talking to Maksym, he thinks there should be no longer any blocker for making this SDK swift, let's discuss on Monday to see if it's still possible to implement this in swift.

File: GoogleMulticastAppDelegate.podspec
Comment (paulb777) [reply GoogleMulticastAppDelegate.podspec:2]: Oh, I think it's needed for GoogleAppMeasurement and AdMob, so probably needs to stay out of the Firebase repo.

File: GoogleMulticastAppDelegate.podspec
Comment (paulb777) [reply GoogleMulticastAppDelegate.podspec:30]: We're blocked on using Swift here until the next breaking change release, which is maybe OK?

We also need to confirm that it's ok for any 1P dependents to have a Swift dependency.

File: GoogleMulticastAppDelegate.podspec
Comment (chliangGoogle) [reply GoogleMulticastAppDelegate.podspec:2]: And if it's used by GoogleAppMeasurement, I guess we should keep it Google prefixed?

Comment (paulb777) [reply GoogleMulticastAppDelegate.podspec:2]: Yes. Even though it's long we should keep it `GoogleUtilities` prefixed since we have approval for that, but not generic `Google` prefix.

File: GoogleMulticastAppDelegate.podspec
Comment (ncooke3) [reply GoogleMulticastAppDelegate.podspec]: Weighing in: I think the versioning should match GoogleUtilities. Mainly because I think having two GoogleUtilities-related versions to track could be error-prone.

File: GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m
Comment (ncooke3) [inline GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m:56]: I'm curious, why is `CFRetain` needed here? (All I know is that it increases the retain count of the object you pass into it. I don't know it's use cases in an ARC context.)

File: GoogleMulticastAppDelegate/Sources/Public/GoogleUtilities/GULMulticastAppDelegate.h
Comment (ncooke3) [inline GoogleMulticastAppDelegate/Sources/Public/GoogleUtilities/GULMulticastAppDelegate.h:25]: IIUC, the passed in delegate will act like an "interceptor", correct? 

nit: If so, I feel like the parameter name should be something like "delegate" or "interceptingDelegate"... wdyt?

File: GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m
Comment (ncooke3) [inline GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m:82]: Is there a case where `_interceptors` contains a `id<GULApplicationDelegate>` that does **not** respond to the given `aSelector`?

**Review by ncooke3 — COMMENTED** [review]
Nice! Leaving some questions and suggestions:

File: GoogleMulticastAppDelegate/Sources/Public/GoogleUtilities/GULMulticastAppDelegate.h
Comment (chliangGoogle) [reply GoogleMulticastAppDelegate/Sources/Public/GoogleUtilities/GULMulticastAppDelegate.h:31]: you mean the GULMulticastAppDelegate and defaultAppDelegate?

it should be weak because it is passed from test app app delegate, so test app should own the app delegate.

File: GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m
Comment (chliangGoogle) [reply GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m:28]: I also believe alloc] init] is preferred as I did my objective c readability.

File: GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m
Comment (chliangGoogle) [reply GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m:56]: I'm mimicing maksym's original implementation listed in the description.

File: GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m
Comment (chliangGoogle) [reply GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m:82]: yeah because there are multiple interceptors from auth, messaging and test app and each has different list of selectors.

File: GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m
Comment (chliangGoogle) [reply GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m:37]: Submit suggested change through terminal.

File: GoogleMulticastAppDelegate/Sources/Public/GoogleUtilities/GULMulticastAppDelegate.h
Comment (chliangGoogle) [reply GoogleMulticastAppDelegate/Sources/Public/GoogleUtilities/GULMulticastAppDelegate.h]: yeah this should actually be interceptor based on the API review. Google catch.

File: GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m
Comment (chliangGoogle) [reply GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m]: this is the .m file tho, I thought it's only needed for public API for .h file.

File: GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m
Comment (chliangGoogle) [reply GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m]: Good catch. Done!

File: GoogleMulticastAppDelegate/Tests/Unit/GULMulticastAppDelegateTest.m
Comment (chliangGoogle) [reply GoogleMulticastAppDelegate/Tests/Unit/GULMulticastAppDelegateTest.m]: Done!. Thanks!

File: GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m
Comment (ncooke3) [reply GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m]: You're right, my bad.

File: GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m
Comment (ncooke3) [reply GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m]: I see now, thanks

File: GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m
Comment (ncooke3) [reply GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m]: Oh, interesting, I did not know that.

File: GoogleMulticastAppDelegate/Sources/Public/GoogleUtilities/GULMulticastAppDelegate.h
Comment (ncooke3) [reply GoogleMulticastAppDelegate/Sources/Public/GoogleUtilities/GULMulticastAppDelegate.h]: Yes, I meant that. Alright, thanks, I understand better.

So, should the property be declared like then:
```diff
- @property(nonatomic, copy) id<GULApplicationDelegate> defaultAppDelegate;
+ @property(nonatomic, weak, nullable) id<GULApplicationDelegate> defaultAppDelegate;
```

Comment (chliangGoogle) [reply GoogleMulticastAppDelegate/Sources/Public/GoogleUtilities/GULMulticastAppDelegate.h]: Just realize we need to retain it to keep it in the interceptors list so I think this should be copy.

File: GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m
Comment (chliangGoogle) [reply GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m]: It's easier to manage and better control for read and write. There might be more reasons but I prefer this way to ensure I have full control of it.

Comment (ncooke3) [reply GoogleMulticastAppDelegate/Sources/GULMulticastAppDelegate.m]: thanks for the explanation– I'm going to try this 😄

**Review by ncooke3 — APPROVED** [review]

**Review by paulb777 — COMMENTED** [review]
Perhaps, this should be merged to a branch instead of main.

I see at least four outstanding issues
- Still needs a Package.swift update
- Multi-platform support
- Integration testing from a podspec deployed to SpecsTesting
- test specs

**Comment by chliangGoogle** [comment]
Yep, will do a series of more PRs to the feature branch so this one won't get large.

**Pull request merged**
